### PR TITLE
[Experimental] hide/show logic improvement

### PR DESCRIPTION
**Pull request by mcmonkey4eva**
for #470

Experimental tweak to track both isInVisibleNodes (ie on/off screen boundary), and shouldHide (based on scale and all) as dataset values (I'd store it on `widget`, but it appears `DOMWidget` is used for one and `LiteGraph.IWidget` is the other, which is awkward - not sure if there's a convenient way to grab the DOMWidget from a litegraph one?)

Per the description of the issue in #470 , the actual style.display value was toggling rapidly before - this change should effectively prevent that.
While editing this, I also added a change to only fire `onHide` if it's actually being hidden in that frame (as opposed to every frame while it's hidden)

There's potentially a rootmore issue that a lot of redundant calculations are being done to check the visibility state that could be skipped? Not sure if `draw` is being called when hidden but I think not, `computeVisibleNodes` could probably be simplified to not subloop if the visible state hasn't changed

Will ask #470 author to test before this is pulled

### CODE REVIEW

**Comment by christian-byrne** [comment]
Nice, massive improvement:

Selection_260

Per your comment on the original issue, setting `display` to `undefined` seems to just set it to the string "undefined," preventing the browser from applying any optimizations. If I update the other method to use `null` like how you did here, it improves even more:

Selection_264

**Comment by huchenlei** [comment]
https://github.com/user-attachments/assets/96691896-114a-406f-be92-5c4488243240

We should probably add toggle of node that has DOM widget to playwright test.

**Review by huchenlei — CHANGES_REQUESTED** [review]
Please add 
- a playwright test to collapse/minimize a node with DOM widget, and then expand.
- a playwright test to zoom out enough so that DOM wigets are hidden, and zoom back.
- a playwright test to drag the canvas so the DOM widget out of view, and drag back.

**Comment by mcmonkey4eva** [comment]
So the root issue this failed is: the pre-existing code was wrong, but worked by luck

The check `node.flags?.collapsed` never passes: it is `false` when open, and `undefined` when closed. It momentarily reaches a `true` state inside the `collapse` function but this doesn't persist into the draw function.
but the collapse handling function farther down sets the element to hidden manually in the func, which makes it seem to work.
This would immediately lose state and fail, however thanks to a magical comedy of errors in perfect sequence, the `undefined` in the computeVisibleNodes glitches the state enough that it ends up kinda just working out at the end.

tldr: none of this code ever worked, but the end result did anyway because it was all broken in exactly the right way to fit together.

I pushed a commit to this PR that now actually persists a `collapsed` state. (Arguably, the flag value should be looked into re why it's wrong?)

Please merge https://github.com/Comfy-Org/ComfyUI_frontend/pull/540 before this PR, as that adds the requested Playwright tests to validate things work as intended.

**Review by huchenlei — APPROVED** [review]
